{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d9b66b4c",
   "metadata": {},
   "source": [
    "## Hybrid de novo protein assembly workflow\n",
    "\n",
    "The hybrid assembly workflow considers contigs obtained from 2 assembly methods: greedy and DBG."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57ba8c25",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "\n",
    "script_dir = os.getcwd()\n",
    "sys.path.append(os.path.join(script_dir, \"../src/instanexus\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b9dafa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my modules\n",
    "import dbg\n",
    "import greedy_method as greedy\n",
    "import mapping as map\n",
    "import consensus as cons\n",
    "import alignment as align\n",
    "import clustering as clus\n",
    "import preprocessing as prep\n",
    "import compute_statistics as comp_stat\n",
    "#import model_peptide_selector as selector\n",
    "\n",
    "# import libraries\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "\n",
    "#import joblib\n",
    "import json\n",
    "import Bio\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8645e16f",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    # works if you are in a script: __file__ exists\n",
    "    BASE_DIR = Path(__file__).resolve().parents[2]\n",
    "except NameError:\n",
    "    # works if you are in a notebook: __file__ does not exist\n",
    "    BASE_DIR = Path().resolve()\n",
    "    # go up until the project folder\n",
    "    while BASE_DIR.name != \"InstaNexus\" and BASE_DIR != BASE_DIR.parent:\n",
    "        BASE_DIR = BASE_DIR.parent\n",
    "\n",
    "JSON_DIR = BASE_DIR / \"json\"\n",
    "INPUT_DIR = BASE_DIR / \"inputs\"\n",
    "FASTA_DIR = BASE_DIR / \"fasta\"\n",
    "OUTPUTS_DIR = BASE_DIR / \"outputs\"\n",
    "FIGURES_DIR = BASE_DIR / \"figures\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3287f4a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sample_metadata(run, chain=\"\", json_path=JSON_DIR / \"sample_metadata.json\"):\n",
    "    with open(json_path, \"r\") as f:\n",
    "        all_meta = json.load(f)\n",
    "\n",
    "    if run not in all_meta:\n",
    "        raise ValueError(f\"Run '{run}' not found in metadata.\")\n",
    "\n",
    "    entries = all_meta[run]\n",
    "\n",
    "    for entry in entries:\n",
    "        if entry[\"chain\"] == chain:\n",
    "            return entry\n",
    "\n",
    "    raise ValueError(f\"No metadata found for run '{run}' with chain '{chain}'.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e967e37",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_colors_from_run(cat, is_scaffold=False, json_path=JSON_DIR / \"colors.json\"):\n",
    "    if not os.path.exists(json_path):\n",
    "        raise FileNotFoundError(f\"Missing color file: {json_path}\")\n",
    "\n",
    "    with open(json_path, \"r\") as f:\n",
    "        colors = json.load(f)\n",
    "\n",
    "    category = cat.split(\"_\")[0].lower()\n",
    "    key = \"scaffold\" if is_scaffold else \"contig\"\n",
    "\n",
    "    try:\n",
    "        return colors[category][key]\n",
    "    except KeyError:\n",
    "        raise ValueError(f\"Color not defined for category '{category}' and key '{key}'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccaf9cbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_combination_name(\n",
    "    ass_method,\n",
    "    conf,\n",
    "    kmer_size,\n",
    "    size_threshold,\n",
    "    min_overlap\n",
    "):\n",
    "    if ass_method in (\"dbg\", \"hybrid\"):\n",
    "        return f\"comb_{ass_method}_c{conf}_ks{kmer_size}_ts{size_threshold}_mo{min_overlap}\"\n",
    "    \n",
    "    elif ass_method == \"greedy\":\n",
    "        return f\"comb_{ass_method}_c{conf}_ts{size_threshold}_mo{min_overlap}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de65b29e",
   "metadata": {},
   "outputs": [],
   "source": [
    "run = \"ma1\"\n",
    "chain = \"light\"\n",
    "\n",
    "meta = get_sample_metadata(run, chain)\n",
    "\n",
    "protein = meta[\"protein\"]\n",
    "chain = meta[\"chain\"]\n",
    "proteases = meta[\"proteases\"]\n",
    "\n",
    "print(chain)\n",
    "print(proteases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf1b0f45",
   "metadata": {},
   "outputs": [],
   "source": [
    "ass_method = \"hybrid\"\n",
    "kmer_size = 6\n",
    "conf = 0.95\n",
    "size_threshold = 10\n",
    "min_overlap = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efe91ca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb = get_combination_name(\n",
    "    ass_method,\n",
    "    conf,\n",
    "    kmer_size,\n",
    "    size_threshold,\n",
    "    min_overlap)\n",
    "\n",
    "print(comb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3461c687",
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\n",
    "    \"ass_method\": ass_method,\n",
    "    \"conf\": conf,\n",
    "    \"size_threshold\": size_threshold,\n",
    "    \"min_overlap\": min_overlap\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02a6bcdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_outputs = OUTPUTS_DIR / f\"{run}{chain}\"\n",
    "\n",
    "prep.create_directory(folder_outputs)\n",
    "\n",
    "combination_folder_out = os.path.join(\n",
    "    folder_outputs,\n",
    "    f\"comb_{ass_method}_c{conf}_ks{kmer_size}_ts{size_threshold}_mo{min_overlap}\",\n",
    ")\n",
    "\n",
    "prep.create_subdirectories_outputs(combination_folder_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cea5b8cd",
   "metadata": {},
   "source": [
    "### Data cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88371e65",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_norm = prep.normalize_sequence(protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44d02d5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(INPUT_DIR / f\"{run}.csv\")\n",
    "\n",
    "df[\"protease\"] = df[\"experiment_name\"].apply(\n",
    "    lambda name: prep.extract_protease(name, proteases)\n",
    ")\n",
    "\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ac3a6b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.protease.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01880cb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = prep.clean_dataframe(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1d96795",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d6f9a9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"cleaned_preds\"] = df[\"preds\"].apply(prep.remove_modifications)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "509283bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ac871a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1945e3e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_contaminants_df(df, seq_col, run, contaminants_fasta):\n",
    "\n",
    "    contam_records = []\n",
    "    for record in SeqIO.parse(contaminants_fasta, \"fasta\"):\n",
    "        if run == \"bsa\" and \"Bovine serum albumin precursor\" in record.description:\n",
    "            continue\n",
    "        contam_records.append(str(record.seq))\n",
    "\n",
    "    mask = ~df[seq_col].apply(\n",
    "        lambda seq: any(seq in contam_seq for contam_seq in contam_records)\n",
    "    )\n",
    "\n",
    "    filtered_df = df[mask].copy()\n",
    "    removed_count = (~mask).sum()\n",
    "\n",
    "    print(f\"Removed {removed_count} contaminant sequences, {len(filtered_df)} remaining.\")\n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "329a7933",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = filter_contaminants_df(df, \"cleaned_preds\", run, FASTA_DIR / \"contaminants.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1aab00d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aac0d21f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42ca0cf9",
   "metadata": {},
   "source": [
    "### Comments\n",
    "\n",
    "A reference free approach assume you do not have any reference protein.\\\n",
    "In this scenario it is not possible to know when the PSM coverage is 100%.\\\n",
    "Having high PSM coverage would allow a graph based approach to work properly.\\ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52b6e813",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(conf)\n",
    "\n",
    "df = df[df[\"conf\"] > conf]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "379b2d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_preds = df[\"cleaned_preds\"].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90da2497",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_preds[:5]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22992f6f",
   "metadata": {},
   "source": [
    "### Hybrid assembly\n",
    "\n",
    "#### Greedy assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5af84e51",
   "metadata": {},
   "outputs": [],
   "source": [
    "greedy_contigs = greedy.assemble_contigs(filtered_preds, min_overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f62a4b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "assembled_contigs = list(set(greedy_contigs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe0cc69",
   "metadata": {},
   "outputs": [],
   "source": [
    "assembled_contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a4d3bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "assembled_contigs = sorted(assembled_contigs, key=len, reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31777d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(assembled_contigs[:5])\n",
    "print(len(assembled_contigs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50de3cad",
   "metadata": {},
   "outputs": [],
   "source": [
    "greedy_scaffolds = greedy.scaffold_iterative_greedy(assembled_contigs,\n",
    "                                                   min_overlap,\n",
    "                                                   size_threshold)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc350462",
   "metadata": {},
   "outputs": [],
   "source": [
    "greedy_scaffolds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b9d222b",
   "metadata": {},
   "source": [
    "### DBG assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02f37d1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "kmers = dbg.get_kmers(assembled_contigs, kmer_size=9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41d76e3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(kmers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bef4c14",
   "metadata": {},
   "outputs": [],
   "source": [
    "edges = dbg.get_debruijn_edges_from_kmers(kmers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07a5712d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbg_contigs = dbg.assemble_contigs(edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de6a2ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbg_contigs = sorted(dbg_contigs, key=len, reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79a9152a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbg_contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecb1c433",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbg_contigs = list(set(dbg_contigs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c30de12",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbg_contigs = [seq for seq in dbg_contigs if len(seq) > size_threshold]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad5be998",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbg_contigs = sorted(dbg_contigs, key=len, reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e58b7234",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbg_contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96737ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbg_scaffolds = dbg.scaffold_iterative(dbg_contigs, min_overlap, size_threshold)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f63bdde2",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_scaffolds = dbg_scaffolds + greedy_scaffolds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b06561d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(all_scaffolds))\n",
    "print(len(dbg_scaffolds))\n",
    "print(len(greedy_scaffolds))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ce94987",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapped_scaffolds = map.process_protein_contigs_scaffold(\n",
    "    all_scaffolds, protein_norm, max_mismatches = 0, min_identity = 0.90\n",
    ")\n",
    "\n",
    "map.mapping_substitutions(mapped_scaffolds, protein_norm, title= \"scaffolds mapped in RF-selected peptides\")"
   ]
  },
  {
   "cell_type": "raw",
   "id": "809e82d4",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "map.mapping_substitutions(\n",
    "    mapped_sequences=mapped_scaffolds,\n",
    "    prot_seq=protein_norm,\n",
    "    title=f\"Contig mapping to reference sequence, {run}\",\n",
    "    contig_colors=get_colors_from_run(\"antibodies\", is_scaffold=False),\n",
    "    match_color=get_colors_from_run(\"antibodies\", is_scaffold=False),\n",
    "    output_file=FIGURES_DIR / f\"fig_X_{run}_substitution_map_contigs_dbg.svg\",\n",
    "    output_folder=\".\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e83a187e",
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "\n",
    "for i, seq in enumerate(all_scaffolds):\n",
    "    record = Bio.SeqRecord.SeqRecord(\n",
    "        Bio.Seq.Seq(seq), id=f\"scaffold_{i+1}\", description=f\"length: {len(seq)}\"\n",
    "    )\n",
    "    records.append(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba417995",
   "metadata": {},
   "outputs": [],
   "source": [
    "Bio.SeqIO.write(\n",
    "    records,\n",
    "    f\"{combination_folder_out}/scaffolds/scaffolds.fasta\",\n",
    "    \"fasta\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca2cd3ab",
   "metadata": {},
   "source": [
    "## Postprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbc2d0d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_id = f\"{ass_method}_{conf}_{run}\"\n",
    "scaffolds_folder_out = Path(f\"{combination_folder_out}/scaffolds\")\n",
    "clustering_out = scaffolds_folder_out / \"clustering\"\n",
    "alignment_out = scaffolds_folder_out / \"alignment\"\n",
    "consensus_out = scaffolds_folder_out / \"consensus\"\n",
    "\n",
    "clustering_out.mkdir(parents=True, exist_ok=True)\n",
    "alignment_out.mkdir(parents=True, exist_ok=True)\n",
    "consensus_out.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c54868d1",
   "metadata": {},
   "source": [
    "### Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d36e9730",
   "metadata": {},
   "outputs": [],
   "source": [
    "clus.cluster_fasta_files(input_folder=str(scaffolds_folder_out))\n",
    "\n",
    "fasta_input = scaffolds_folder_out / f\"scaffolds.fasta\"\n",
    "\n",
    "cluster_tsv_folder = clustering_out / run_id\n",
    "    \n",
    "clus.process_fasta_and_clusters(\n",
    "        fasta_file=str(fasta_input),\n",
    "        input_folder=str(scaffolds_folder_out),\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dad9b38",
   "metadata": {},
   "source": [
    "### Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a3e3ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "align.process_alignment(input_folder=str(scaffolds_folder_out))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "695120a2",
   "metadata": {},
   "source": [
    "### Consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "346c599d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cons.process_alignment_files(\n",
    "        align_folder=str(alignment_out),\n",
    "        output_folder=str(consensus_out),\n",
    "        run_id=run_id,\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "instanexus",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
