diff --git a/ldastoolsal/src/reverse.hh.in b/ldastoolsal/src/reverse.hh.in
index 8676e0ef4..2ac59b097 100644
--- a/ldastoolsal/src/reverse.hh.in
+++ b/ldastoolsal/src/reverse.hh.in
@@ -34,8 +34,6 @@
 #define bswap_16( x ) bswap16( x )
 #elif @HAVE_BSWAP_16_MACRO@ // HAVE_BSWAP_16_MACRO
 #define bswap_16( x ) BSWAP_16( x )
-#elif @HAVE__OSSWAPINT16@ // HAVE__OSSWAPINT16
-#define bswap_16( x ) _OSSwapInt16( x )
 #elif @HAVE_OSSWAPINT16@ // HAVE_OSSWAPINT16
 #define bswap_16( x ) OSSwapInt16( x )
 #else
@@ -48,8 +46,6 @@
 #define bswap_32( x ) bswap32( x )
 #elif @HAVE_BSWAP_32_MACRO@ // HAVE_BSWAP_32_MACRO
 #define bswap_32( x ) BSWAP_32( x )
-#elif @HAVE__OSSWAPINT32@ // HAVE__OSSWAPINT32
-#define bswap_32( x ) _OSSwapInt32( x )
 #elif @HAVE_OSSWAPINT32@ // HAVE_OSSWAPINT32
 #define bswap_32( x ) OSSwapInt32( x )
 #else
@@ -62,8 +58,6 @@
 #define bswap_64( x ) bswap64( x )
 #elif @HAVE_BSWAP_64_MACRO@ // HAVE_BSWAP_64_MACRO
 #define bswap_64( x ) BSWAP_64( x )
-#elif @HAVE__OSSWAPINT64@ // HAVE__OSSWAPINT64
-#define bswap_64( x ) _OSSwapInt64( x )
 #elif @HAVE_OSSWAPINT64@ // HAVE_OSSWAPINT64
 #define bswap_64( x ) OSSwapInt64( x )
 #else
