diff --git a/ldastoolsal/src/CMakeLists.txt b/ldastoolsal/src/CMakeLists.txt
index 1a99a2ab8..be8dea95f 100644
--- a/ldastoolsal/src/CMakeLists.txt
+++ b/ldastoolsal/src/CMakeLists.txt
@@ -211,7 +211,7 @@ cx_link_headers( ${HDR_DIR} ${ldastoolsalhh_HEADERS} )
 cx_link_headers( BUILT_HEADERS ${HDR_DIR} ${nodist_ldastoolsalhh_HEADERS} )

 set( libldastoolsal_la_LIBADD
-  ${Boost_LOG_LIBRARY} ${Boost_LIBRARIES} stdc++fs ${CMAKE_DL_LIBS}
+  ${Boost_LOG_LIBRARY} ${Boost_LIBRARIES} ${CMAKE_DL_LIBS}
   )

 cx_target_library(
diff --git a/ldastoolsal/src/ldastoolsal.pc.in b/ldastoolsal/src/ldastoolsal.pc.in
index 80137cbe6..b825b633a 100644
--- a/ldastoolsal/src/ldastoolsal.pc.in
+++ b/ldastoolsal/src/ldastoolsal.pc.in
@@ -10,5 +10,5 @@ CXXSTDLDFLAGS=@CXXSTDLDFLAGS@
 Name: ldastoolsal
 Description: Library which implements abstraction layer for the ldas-tools package.
 Version: @PROJECT_VERSION@
-Libs: -L${libdir}  -lldastoolsal ${CXXSTDLDFLAGS} @Boost_LIBRARIES@ stdc++fs
+Libs: -L${libdir}  -lldastoolsal ${CXXSTDLDFLAGS} @Boost_LIBRARIES@
 Cflags: ${CXXSTDFLAGS} -I${includedir} @BOOST_CPPFLAGS@
