{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef1f0227",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/chlab/flync/.venv/lib/python3.12/site-packages/sorted_nearest/__init__.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8ddfa0c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dm6 = pr.read_gtf('Drosophila_melanogaster.BDGP6.46.113.gtf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2b504e76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the data for cannonical chromosomes\n",
    "chromosomes = ['2L', '2R', '3L', '3R', '4', 'X', 'Y']\n",
    "\n",
    "# Filter the data for the specified chromosomes\n",
    "dm6 = dm6[dm6.Chromosome.isin(chromosomes)]\n",
    "\n",
    "# Define positive and negative biotypes. Positive `y=True` and negative `y=False` for classification\n",
    "pos_biotype = 'ncRNA'\n",
    "neg_biotype = 'protein_coding'\n",
    "\n",
    "# Pandas API selectors\n",
    "# Split by positive and negative biotypes ( genes OR transcripts )\n",
    "ncrna_genes = ((dm6.gene_biotype == pos_biotype))\n",
    "pc_genes = ((dm6.gene_biotype == neg_biotype))\n",
    "\n",
    "# Filter the data for the specified positive and negative biotypes\n",
    "ncr = dm6[ncrna_genes]\n",
    "pcg = dm6[pc_genes]\n",
    "\n",
    "# Select the features of interest (transcripts & exons) and filter by biotype\n",
    "\n",
    "# To get lncRNA, filter transcripts by length >= 200bp\n",
    "lncrna_transcripts = ((ncr.Feature == 'transcript') & (ncr.transcript_biotype == pos_biotype) & (ncr.End - ncr.Start > 200))\n",
    "# Select exons that are part of the lncRNA transcripts\n",
    "lncrna_exons = ((ncr.Feature == 'exon') & (ncr.transcript_id.isin(ncr[lncrna_transcripts].transcript_id)))\n",
    "\n",
    "# Get protein coding transcripts\n",
    "pcg_transcripts = ((pcg.Feature == 'transcript') & (pcg.transcript_biotype == neg_biotype) & (pcg.End - pcg.Start > 10))\n",
    "# Select exons that are part of the protein coding transcripts\n",
    "pcg_exons = ((pcg.Feature == 'exon') & (pcg.transcript_id.isin(pcg[pcg_transcripts].transcript_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9cebaf14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Keep all filtered transcripts & exons\n",
    "ncr_transcripts = ncr[ (lncrna_transcripts) ]\n",
    "init_ncr_count = len(ncr_transcripts)\n",
    "# Remove duplicates based on the chromosome, start, and end positions\n",
    "ncr_transcripts = ncr_transcripts[~(ncr_transcripts.df.duplicated(subset=['Chromosome', 'Start', 'End']))]\n",
    "dedup_ncr_count = len(ncr_transcripts)\n",
    "\n",
    "pcg_transcripts = pcg[ (pcg_transcripts) ]\n",
    "init_pcg_count = len(pcg_transcripts)\n",
    "pcg_transcripts = pcg_transcripts[~(pcg_transcripts.df.duplicated(subset=['Chromosome', 'Start', 'End']))]\n",
    "dedup_pcg_count = len(pcg_transcripts)\n",
    "\n",
    "ncr_exons = ncr [ (lncrna_exons) ]\n",
    "init_ncr_count += len(ncr_exons)\n",
    "ncr_exons = ncr_exons[~(ncr_exons.df.duplicated(subset=['Chromosome', 'Start', 'End', 'transcript_id']))]\n",
    "dedup_ncr_count += len(ncr_exons)\n",
    "pcg_exons = pcg [ (pcg_exons) ]\n",
    "init_pcg_count += len(pcg_exons)\n",
    "pcg_exons = pcg_exons[~(pcg_exons.df.duplicated(subset=['Chromosome', 'Start', 'End', 'transcript_id']))]\n",
    "dedup_pcg_count += len(pcg_exons)\n",
    "\n",
    "ncr_base = pr.concat([ncr_transcripts, ncr_exons])\n",
    "pcg_base = pr.concat([pcg_transcripts, pcg_exons])\n",
    "\n",
    "# Deduplicate the data by removing duplicates based on the chromosome, start, and end positions\n",
    "# When duplicated (transcript = exon) keep exons `keep='last'` as exons appear after transcripts in the GTF\n",
    "# ncr_base = ncr_base[~(ncr_base.df.duplicated(subset=['Chromosome', 'Start', 'End'], keep=\"last\"))]\n",
    "# pcg_base = pcg_base[~(pcg_base.df.duplicated(subset=['Chromosome', 'Start', 'End'], keep=\"last\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7bd6bd57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lncRNA transcripts: initial count = 7901, deduplicated count = 7708\n",
      "protein coding transcripts: initial count = 214306, deduplicated count = 206955\n",
      "Total removed lncRNA entries: 193 | Percentage: 2.44%\n",
      "Total removed protein coding entries: 7351 | Percentage: 3.43%\n"
     ]
    }
   ],
   "source": [
    "# Output the difference in counts before and after deduplication\n",
    "print(f\"lncRNA transcripts: initial count = {init_ncr_count}, deduplicated count = {dedup_ncr_count}\")\n",
    "print(f\"protein coding transcripts: initial count = {init_pcg_count}, deduplicated count = {dedup_pcg_count}\")\n",
    "print(f\"Total removed lncRNA entries: {init_ncr_count - dedup_ncr_count} | Percentage: {((init_ncr_count - dedup_ncr_count)/init_ncr_count)*100:.2f}%\")\n",
    "print(f\"Total removed protein coding entries: {init_pcg_count - dedup_pcg_count} | Percentage: {((init_pcg_count - dedup_pcg_count)/init_pcg_count)*100:.2f}%\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cb6df271",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((7708, 20), (206955, 20))"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncr_base.df.shape, pcg_base.df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d259963b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((2802, 20), (23419, 20))"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncr_base.df[ncr_base.df.Feature == 'transcript'].shape, pcg_base.df[pcg_base.df.Feature == 'transcript'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "012537a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the filtered data to GTF files\n",
    "ncr_base.to_gtf('ncr_base.gtf')\n",
    "pcg_base.to_gtf('pcg_base.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4e7e2929",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save BED files\n",
    "ncr_bed = ncr_base.to_bed('ncr_base.bed', keep=False, chain=True)\n",
    "pcg_bed = pcg_base.to_bed('pcg_base.bed', keep=False, chain=True)\n",
    "\n",
    "ncr_bed.Name = ncr_base.transcript_id\n",
    "pcg_bed.Name = pcg_base.transcript_id\n",
    "\n",
    "ncr_bed.to_bed('ncr_base.bed', keep=False)\n",
    "pcg_bed.to_bed('pcg_base.bed', keep=False)\n",
    "\n",
    "# Save parquet files\n",
    "ncr_base.df.to_parquet('ncr_base.parquet', index=False)\n",
    "pcg_base.df.to_parquet('pcg_base.parquet', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "flync",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
