{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "00c2b272",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[2mUsing Python 3.12.9 environment at: /home/chlab/flync/.venv\u001b[0m\n",
      "\u001b[2mResolved \u001b[1m120 packages\u001b[0m \u001b[2min 185ms\u001b[0m\u001b[0m\n",
      "   \u001b[36m\u001b[1mBuilding\u001b[0m\u001b[39m flync\u001b[2m @ file:///home/chlab/flync\u001b[0m\n",
      "\u001b[2mResolved \u001b[1m120 packages\u001b[0m \u001b[2min 185ms\u001b[0m\u001b[0m\n",
      "   \u001b[36m\u001b[1mBuilding\u001b[0m\u001b[39m flync\u001b[2m @ file:///home/chlab/flync\u001b[0m\n",
      "      \u001b[32m\u001b[1mBuilt\u001b[0m\u001b[39m flync\u001b[2m @ file:///home/chlab/flync\u001b[0m\n",
      "\u001b[2mPrepared \u001b[1m1 package\u001b[0m \u001b[2min 421ms\u001b[0m\u001b[0m\n",
      "\u001b[2mUninstalled \u001b[1m1 package\u001b[0m \u001b[2min 0.35ms\u001b[0m\u001b[0m\n",
      "\u001b[2mInstalled \u001b[1m1 package\u001b[0m \u001b[2min 0.71ms\u001b[0m\u001b[0m\n",
      " \u001b[33m~\u001b[39m \u001b[1mflync\u001b[0m\u001b[2m==1.0.0 (from file:///home/chlab/flync)\u001b[0m\n",
      "      \u001b[32m\u001b[1mBuilt\u001b[0m\u001b[39m flync\u001b[2m @ file:///home/chlab/flync\u001b[0m\n",
      "\u001b[2mPrepared \u001b[1m1 package\u001b[0m \u001b[2min 421ms\u001b[0m\u001b[0m\n",
      "\u001b[2mUninstalled \u001b[1m1 package\u001b[0m \u001b[2min 0.35ms\u001b[0m\u001b[0m\n",
      "\u001b[2mInstalled \u001b[1m1 package\u001b[0m \u001b[2min 0.71ms\u001b[0m\u001b[0m\n",
      " \u001b[33m~\u001b[39m \u001b[1mflync\u001b[0m\u001b[2m==1.0.0 (from file:///home/chlab/flync)\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "uv pip install /home/chlab/flync --upgrade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f09bd559",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import tempfile\n",
    "from flync.features.feature_wrapper import FeatureWrapper as fw\n",
    "import flync.features.feature_cleaning as ct\n",
    "\n",
    "# Create a temporary directory and return its path\n",
    "# Use mkdtemp() so tmp_dir is a string path usable across notebook cells\n",
    "tmp_dir = tempfile.mkdtemp()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4cdaebcb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-31 17:07:39,297 - INFO - feature_wrapper - Using base cache directory: /tmp/tmp0e5ryw87\n",
      "2025-10-31 17:07:39,298 - INFO - feature_wrapper - Initialized FeatureWrapper\n",
      "2025-10-31 17:07:39,298 - INFO - feature_wrapper - Initialized FeatureWrapper\n"
     ]
    }
   ],
   "source": [
    "fw = fw(\n",
    "    cache_dir=tmp_dir,\n",
    "    threads=20\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e178d7ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-31 17:14:32,157 - INFO - feature_wrapper - Running all feature extraction scripts...\n",
      "Feature extraction:   0%|          | 0/3 [00:00<?, ?modules/s]2025-10-31 17:14:32,159 - INFO - root - Starting Feature extraction (3 items)\n",
      "2025-10-31 17:14:32,160 - INFO - feature_wrapper - Converting GTF to BED: /home/chlab/flync/src/data/ncr_base.gtf -> /tmp/tmpcuq0ti_7/transcripts_from_gtf.bed\n",
      "2025-10-31 17:14:32,234 - INFO - feature_wrapper - Loaded GTF with 7708 total features\n",
      "2025-10-31 17:14:32,243 - INFO - feature_wrapper - Found 2802 transcript features\n",
      "2025-10-31 17:14:32,250 - INFO - feature_wrapper - Wrote 2802 transcripts to BED file: /tmp/tmpcuq0ti_7/transcripts_from_gtf.bed\n",
      "2025-10-31 17:14:32,251 - INFO - feature_wrapper - Generated BED from GTF: /tmp/tmpcuq0ti_7/transcripts_from_gtf.bed\n",
      "Running BigWig Query:   0%|          | 0/3 [00:00<?, ?modules/s]2025-10-31 17:14:32,252 - INFO - feature_wrapper - Running BigWig Query feature extraction...\n",
      "2025-10-31 17:14:32,252 - INFO - feature_wrapper - Using configuration file: /home/chlab/flync/src/config/bwq_config.yaml\n",
      "2025-10-31 17:14:32,253 - INFO - root - Loading configuration from /home/chlab/flync/src/config/bwq_config.yaml\n",
      "2025-10-31 17:14:32,257 - INFO - root - Loaded 26 stat/extraction configurations from /home/chlab/flync/src/config/bwq_config.yaml\n",
      "2025-10-31 17:14:32,257 - INFO - root - Reading genomic ranges from /tmp/tmpcuq0ti_7/transcripts_from_gtf.bed\n",
      "2025-10-31 17:14:32,297 - INFO - root - Extracted 2802 ranges from /tmp/tmpcuq0ti_7/transcripts_from_gtf.bed using pyranges.\n",
      "2025-10-31 17:14:32,297 - INFO - root - Processing queries with 20 workers\n",
      "2025-10-31 17:14:32,298 - INFO - root - Using specified max_workers: 20\n",
      "2025-10-31 17:14:32,299 - INFO - root - Preloading 10 unique file paths...\n",
      "2025-10-31 17:14:32,300 - INFO - root - Starting Downloading/caching files (10 items)\n",
      "2025-10-31 17:14:32,301 - INFO - root - Attempting to preload 10 remote files...\n",
      "2025-10-31 17:15:02,314 - WARNING - root - Timeout waiting for http://expdata.cmmt.ubc.ca/JASPAR/downloads/UCSC_tracks/2022/JASPAR2022_dm6.bb download from another thread. This thread will attempt.\n",
      "2025-10-31 17:15:02,315 - WARNING - root - Timeout waiting for https://hgdownload.soe.ucsc.edu/goldenPath/dm6/phastCons27way/dm6.27way.phastCons.bw download from another thread. This thread will attempt.\n",
      "2025-10-31 17:15:02,316 - INFO - root - Downloading https://hgdownload.soe.ucsc.edu/goldenPath/dm6/phastCons27way/dm6.27way.phastCons.bw to persistent cache: /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fhgdownload.soe.ucsc.edu%2FgoldenPath%2Fdm6%2FphastCons27way%2Fdm6.27way.phastCons.bw\n",
      "2025-10-31 17:15:02,315 - WARNING - root - Timeout waiting for https://epd.expasy.org/epd/ucsc/epdHub/dm6/GSM480156_dm3-S2-H3K4me3.bw download from another thread. This thread will attempt.\n",
      "2025-10-31 17:15:02,316 - INFO - root - Downloading http://expdata.cmmt.ubc.ca/JASPAR/downloads/UCSC_tracks/2022/JASPAR2022_dm6.bb to persistent cache: /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_http%3A%2F%2Fexpdata.cmmt.ubc.ca%2FJASPAR%2Fdownloads%2FUCSC_tracks%2F2022%2FJASPAR2022_dm6.bb\n",
      "2025-10-31 17:15:02,315 - WARNING - root - Timeout waiting for https://hgdownload.soe.ucsc.edu/gbdb/dm6/bbi/gc5BaseBw/gc5Base.bw download from another thread. This thread will attempt.\n",
      "2025-10-31 17:15:02,319 - INFO - root - Downloading https://hgdownload.soe.ucsc.edu/gbdb/dm6/bbi/gc5BaseBw/gc5Base.bw to persistent cache: /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fhgdownload.soe.ucsc.edu%2Fgbdb%2Fdm6%2Fbbi%2Fgc5BaseBw%2Fgc5Base.bw\n",
      "2025-10-31 17:15:02,318 - INFO - root - Downloading https://epd.expasy.org/epd/ucsc/epdHub/dm6/GSM480156_dm3-S2-H3K4me3.bw to persistent cache: /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FGSM480156_dm3-S2-H3K4me3.bw\n",
      "2025-10-31 17:15:03,088 - INFO - root - Successfully downloaded https://epd.expasy.org/epd/ucsc/epdHub/dm6/GSM480156_dm3-S2-H3K4me3.bw to /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FGSM480156_dm3-S2-H3K4me3.bw\n",
      "2025-10-31 17:15:03,088 - INFO - root - Successfully opened remote BigWig/BigBed file: https://epd.expasy.org/epd/ucsc/epdHub/dm6/GSM480156_dm3-S2-H3K4me3.bw (from /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FGSM480156_dm3-S2-H3K4me3.bw)\n",
      "2025-10-31 17:15:33,097 - WARNING - root - Timeout waiting for https://hgdownload.soe.ucsc.edu/goldenPath/dm6/phyloP124way/dm6.phyloP124way.bw download from another thread. This thread will attempt.\n",
      "2025-10-31 17:15:33,097 - INFO - root - Downloading https://hgdownload.soe.ucsc.edu/goldenPath/dm6/phyloP124way/dm6.phyloP124way.bw to persistent cache: /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fhgdownload.soe.ucsc.edu%2FgoldenPath%2Fdm6%2FphyloP124way%2Fdm6.phyloP124way.bw\n",
      "2025-10-31 17:15:33,373 - INFO - root - Successfully downloaded https://hgdownload.soe.ucsc.edu/gbdb/dm6/bbi/gc5BaseBw/gc5Base.bw to /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fhgdownload.soe.ucsc.edu%2Fgbdb%2Fdm6%2Fbbi%2Fgc5BaseBw%2Fgc5Base.bw\n",
      "2025-10-31 17:15:33,375 - INFO - root - Successfully opened remote BigWig/BigBed file: https://hgdownload.soe.ucsc.edu/gbdb/dm6/bbi/gc5BaseBw/gc5Base.bw (from /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fhgdownload.soe.ucsc.edu%2Fgbdb%2Fdm6%2Fbbi%2Fgc5BaseBw%2Fgc5Base.bw)\n",
      "2025-10-31 17:16:03,382 - WARNING - root - Timeout waiting for https://remap.univ-amu.fr/storage/public/hubReMap2022/dm6/bigBed/remap2022_all_macs2_dm6_v1_0.bb download from another thread. This thread will attempt.\n",
      "2025-10-31 17:16:03,383 - INFO - root - Downloading https://remap.univ-amu.fr/storage/public/hubReMap2022/dm6/bigBed/remap2022_all_macs2_dm6_v1_0.bb to persistent cache: /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fremap.univ-amu.fr%2Fstorage%2Fpublic%2FhubReMap2022%2Fdm6%2FbigBed%2Fremap2022_all_macs2_dm6_v1_0.bb\n",
      "2025-10-31 17:17:43,242 - INFO - root - Successfully downloaded https://hgdownload.soe.ucsc.edu/goldenPath/dm6/phastCons27way/dm6.27way.phastCons.bw to /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fhgdownload.soe.ucsc.edu%2FgoldenPath%2Fdm6%2FphastCons27way%2Fdm6.27way.phastCons.bw\n",
      "2025-10-31 17:17:43,247 - INFO - root - Successfully opened remote BigWig/BigBed file: https://hgdownload.soe.ucsc.edu/goldenPath/dm6/phastCons27way/dm6.27way.phastCons.bw (from /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fhgdownload.soe.ucsc.edu%2FgoldenPath%2Fdm6%2FphastCons27way%2Fdm6.27way.phastCons.bw)\n",
      "2025-10-31 17:17:43,251 - INFO - root - Downloading/caching files: 25% complete (3/10) - elapsed: 191.0s\n",
      "2025-10-31 17:17:47,694 - INFO - root - Successfully downloaded https://remap.univ-amu.fr/storage/public/hubReMap2022/dm6/bigBed/remap2022_all_macs2_dm6_v1_0.bb to /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fremap.univ-amu.fr%2Fstorage%2Fpublic%2FhubReMap2022%2Fdm6%2FbigBed%2Fremap2022_all_macs2_dm6_v1_0.bb\n",
      "2025-10-31 17:17:47,700 - INFO - root - Successfully opened remote BigWig/BigBed file: https://remap.univ-amu.fr/storage/public/hubReMap2022/dm6/bigBed/remap2022_all_macs2_dm6_v1_0.bb (from /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fremap.univ-amu.fr%2Fstorage%2Fpublic%2FhubReMap2022%2Fdm6%2FbigBed%2Fremap2022_all_macs2_dm6_v1_0.bb)\n",
      "2025-10-31 17:18:13,256 - WARNING - root - Timeout waiting for https://epd.expasy.org/epd/ucsc/epdHub/dm6/GSM480159_dm3-S2-PolII.bw download from another thread. This thread will attempt.\n",
      "2025-10-31 17:18:13,256 - INFO - root - Downloading https://epd.expasy.org/epd/ucsc/epdHub/dm6/GSM480159_dm3-S2-PolII.bw to persistent cache: /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FGSM480159_dm3-S2-PolII.bw\n",
      "2025-10-31 17:18:14,398 - INFO - root - Successfully downloaded https://epd.expasy.org/epd/ucsc/epdHub/dm6/GSM480159_dm3-S2-PolII.bw to /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FGSM480159_dm3-S2-PolII.bw\n",
      "2025-10-31 17:18:14,399 - INFO - root - Successfully opened remote BigWig/BigBed file: https://epd.expasy.org/epd/ucsc/epdHub/dm6/GSM480159_dm3-S2-PolII.bw (from /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FGSM480159_dm3-S2-PolII.bw)\n",
      "2025-10-31 17:18:14,401 - INFO - root - Downloading/caching files: 50% complete (5/10) - elapsed: 222.1s\n",
      "2025-10-31 17:18:17,708 - WARNING - root - Timeout waiting for https://epd.expasy.org/ftp/epdnew/drosophila/005/Dm_EPDnew_005_dm6.bb download from another thread. This thread will attempt.\n",
      "2025-10-31 17:18:17,709 - INFO - root - Downloading https://epd.expasy.org/ftp/epdnew/drosophila/005/Dm_EPDnew_005_dm6.bb to persistent cache: /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fftp%2Fepdnew%2Fdrosophila%2F005%2FDm_EPDnew_005_dm6.bb\n",
      "2025-10-31 17:18:18,001 - INFO - root - Successfully downloaded https://epd.expasy.org/ftp/epdnew/drosophila/005/Dm_EPDnew_005_dm6.bb to /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fftp%2Fepdnew%2Fdrosophila%2F005%2FDm_EPDnew_005_dm6.bb\n",
      "2025-10-31 17:18:18,001 - INFO - root - Successfully opened remote BigWig/BigBed file: https://epd.expasy.org/ftp/epdnew/drosophila/005/Dm_EPDnew_005_dm6.bb (from /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fftp%2Fepdnew%2Fdrosophila%2F005%2FDm_EPDnew_005_dm6.bb)\n",
      "2025-10-31 17:18:44,406 - WARNING - root - Timeout waiting for https://epd.expasy.org/epd/ucsc/epdHub/dm6/All_CAGE_plus.bw download from another thread. This thread will attempt.\n",
      "2025-10-31 17:18:44,406 - INFO - root - Downloading https://epd.expasy.org/epd/ucsc/epdHub/dm6/All_CAGE_plus.bw to persistent cache: /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FAll_CAGE_plus.bw\n",
      "2025-10-31 17:18:47,409 - INFO - root - Successfully downloaded https://epd.expasy.org/epd/ucsc/epdHub/dm6/All_CAGE_plus.bw to /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FAll_CAGE_plus.bw\n",
      "2025-10-31 17:18:47,410 - INFO - root - Successfully opened remote BigWig/BigBed file: https://epd.expasy.org/epd/ucsc/epdHub/dm6/All_CAGE_plus.bw (from /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FAll_CAGE_plus.bw)\n",
      "2025-10-31 17:18:48,009 - WARNING - root - Timeout waiting for https://epd.expasy.org/epd/ucsc/epdHub/dm6/All_CAGE_minus.bw download from another thread. This thread will attempt.\n",
      "2025-10-31 17:18:48,009 - INFO - root - Downloading https://epd.expasy.org/epd/ucsc/epdHub/dm6/All_CAGE_minus.bw to persistent cache: /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FAll_CAGE_minus.bw\n",
      "2025-10-31 17:18:52,906 - INFO - root - Successfully downloaded https://epd.expasy.org/epd/ucsc/epdHub/dm6/All_CAGE_minus.bw to /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FAll_CAGE_minus.bw\n",
      "2025-10-31 17:18:52,907 - INFO - root - Successfully opened remote BigWig/BigBed file: https://epd.expasy.org/epd/ucsc/epdHub/dm6/All_CAGE_minus.bw (from /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fepd.expasy.org%2Fepd%2Fucsc%2FepdHub%2Fdm6%2FAll_CAGE_minus.bw)\n",
      "2025-10-31 17:18:52,908 - INFO - root - Downloading/caching files: 75% complete (8/10) - elapsed: 260.6s\n",
      "2025-10-31 17:19:15,099 - INFO - root - Successfully downloaded https://hgdownload.soe.ucsc.edu/goldenPath/dm6/phyloP124way/dm6.phyloP124way.bw to /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fhgdownload.soe.ucsc.edu%2FgoldenPath%2Fdm6%2FphyloP124way%2Fdm6.phyloP124way.bw\n",
      "2025-10-31 17:19:15,100 - INFO - root - Successfully opened remote BigWig/BigBed file: https://hgdownload.soe.ucsc.edu/goldenPath/dm6/phyloP124way/dm6.phyloP124way.bw (from /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_https%3A%2F%2Fhgdownload.soe.ucsc.edu%2FgoldenPath%2Fdm6%2FphyloP124way%2Fdm6.phyloP124way.bw)\n",
      "2025-10-31 17:34:27,481 - INFO - root - Successfully downloaded http://expdata.cmmt.ubc.ca/JASPAR/downloads/UCSC_tracks/2022/JASPAR2022_dm6.bb to /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_http%3A%2F%2Fexpdata.cmmt.ubc.ca%2FJASPAR%2Fdownloads%2FUCSC_tracks%2F2022%2FJASPAR2022_dm6.bb\n",
      "2025-10-31 17:34:27,488 - INFO - root - Successfully opened remote BigWig/BigBed file: http://expdata.cmmt.ubc.ca/JASPAR/downloads/UCSC_tracks/2022/JASPAR2022_dm6.bb (from /tmp/tmp0e5ryw87/bwq_persistent_cache/bwq_cache_http%3A%2F%2Fexpdata.cmmt.ubc.ca%2FJASPAR%2Fdownloads%2FUCSC_tracks%2F2022%2FJASPAR2022_dm6.bb)\n",
      "2025-10-31 17:34:27,490 - INFO - root - Preloaded 10/10 remote files\n",
      "Downloading/caching files: 100%|██████████| 10/10 [19:55<00:00, 119.52s/it]\n",
      "2025-10-31 17:34:27,491 - INFO - root - Completed Downloading/caching files (10/10 items) - elapsed: 1195.2s\n",
      "2025-10-31 17:34:27,491 - INFO - root - Using ProcessPoolExecutor with 20 workers, batch_size=50\n",
      "2025-10-31 17:34:27,492 - INFO - root - Created 57 batches from 2802 ranges\n",
      "2025-10-31 17:34:27,600 - INFO - root - Submitted 57 batch queries to 20 workers.\n",
      "2025-10-31 17:34:27,613 - INFO - root - Starting Processing 2802 ranges in 57 batches (57 items)\n",
      "2025-10-31 17:34:29,691 - INFO - root - Processing 2802 ranges in 57 batches: 25% complete (15/57) - elapsed: 2.1s\n",
      "2025-10-31 17:34:31,304 - INFO - root - Processing 2802 ranges in 57 batches: 50% complete (29/57) - elapsed: 3.7s\n",
      "2025-10-31 17:34:32,809 - INFO - root - Processing 2802 ranges in 57 batches: 75% complete (43/57) - elapsed: 5.2s\n",
      "Processing 2802 ranges in 57 batches: 100%|██████████| 57/57 [00:11<00:00,  5.02it/s]\n",
      "2025-10-31 17:34:38,960 - INFO - root - Completed Processing 2802 ranges in 57 batches (57/57 items) - elapsed: 11.4s\n",
      "2025-10-31 17:34:38,978 - INFO - root - Collected 2802 results. Creating dataframe.\n",
      "2025-10-31 17:34:38,993 - INFO - feature_wrapper - Saving BigWig Query results to /tmp/tmpcuq0ti_7/bwq_features.parquet\n",
      "Running BigWig Query:  33%|███▎      | 1/3 [20:06<40:13, 1206.88s/modules]2025-10-31 17:34:39,043 - INFO - root - Running BigWig Query: 25% complete (1/3) - elapsed: 1206.9s\n",
      "2025-10-31 17:34:39,043 - INFO - feature_wrapper - BigWig Query completed. Shape: (2802, 30)\n",
      "2025-10-31 17:34:39,043 - INFO - feature_wrapper - Extracting spliced sequences from GTF: /home/chlab/flync/src/data/ncr_base.gtf\n",
      "2025-10-31 17:34:39,159 - INFO - feature_wrapper - Creating gffutils database from GTF...\n",
      "2025-10-31 17:34:39,512 - INFO - feature_wrapper - gffutils database created successfully\n",
      "2025-10-31 17:34:39,513 - INFO - feature_wrapper - Extracting sequences for transcript features...\n",
      "2025-10-31 17:34:52,418 - INFO - feature_wrapper - Processed 2802 transcripts: 2802 written, 0 empty, 0 failed\n",
      "2025-10-31 17:34:52,438 - INFO - feature_wrapper - GTF sequence extraction completed. FASTA: /tmp/tmpcuq0ti_7/sequences_from_gtf.fasta SeqDF shape: (2802, 2)\n",
      "Calculating RNA MFE:  33%|███▎      | 1/3 [20:20<40:13, 1206.88s/modules] 2025-10-31 17:34:52,445 - INFO - feature_wrapper - Running MFE feature extraction...\n",
      "2025-10-31 17:34:52,446 - INFO - root - Calculating MFE for 2802 sequences using 20 processes with LinearFold...\n",
      "2025-10-31 17:34:52,447 - INFO - root - Starting Folding RNA sequences (2,802 items)\n",
      "2025-10-31 17:35:20,246 - INFO - root - Folding RNA sequences: 25% complete (701/2,802) - elapsed: 27.8s\n",
      "2025-10-31 17:35:59,364 - INFO - root - Folding RNA sequences: 50% complete (1,401/2,802) - elapsed: 66.9s\n",
      "2025-10-31 17:36:43,104 - INFO - root - Folding RNA sequences: 75% complete (2,102/2,802) - elapsed: 110.7s\n",
      "Folding RNA sequences: 100%|██████████| 2802/2802 [05:58<00:00,  7.82it/s]\n",
      "2025-10-31 17:40:50,601 - INFO - root - Completed Folding RNA sequences (2,802/2,802 items) - elapsed: 358.2s\n",
      "2025-10-31 17:40:50,601 - INFO - root - MFE calculations complete.\n",
      "2025-10-31 17:40:50,606 - INFO - feature_wrapper - Saving MFE results to /tmp/tmpcuq0ti_7/mfe_features.parquet\n",
      "Calculating RNA MFE:  67%|██████▋   | 2/3 [26:18<11:55, 715.54s/modules] 2025-10-31 17:40:50,649 - INFO - root - Calculating RNA MFE: 50% complete (2/3) - elapsed: 1578.5s\n",
      "2025-10-31 17:40:50,650 - INFO - feature_wrapper - MFE completed. Shape: (2802, 4)\n",
      "Calculating k-mer features:  67%|██████▋   | 2/3 [26:18<11:55, 715.54s/modules]2025-10-31 17:40:50,651 - INFO - feature_wrapper - Running K-mer feature extraction...\n",
      "2025-10-31 17:40:50,655 - INFO - features.kmer - Using 20 worker processes. Batch size: 10,000 sequences.\n",
      "2025-10-31 17:40:50,655 - INFO - features.kmer - Counting BINARY k-mers (A/T->0, G/C->1) for k=3-12.\n",
      "2025-10-31 17:40:50,656 - INFO - root - Generating all possible k-mers for k=3-12 using alphabet '01'...\n",
      "2025-10-31 17:40:50,658 - INFO - root - Generated 8,184 unique k-mers in 0.00s.\n",
      "Using temporary directory for batches: /tmp/tmp0e5ryw87/kmer_temp/kmer_binary_batches_def18359\n",
      "2025-10-31 17:40:50,662 - INFO - root - Starting Processing sequences\n",
      "\n",
      "Processing Batch 1...\n",
      "Reading from file 1/1: sequences_from_gtf.fasta\n",
      "  Read 2802 DNA sequences for batch 1.\n",
      "  Starting binary k-mer counting for 2802 sequences in batch 1...\n",
      "  Building sparse matrix for batch 1...\n",
      "  Saving batch matrix (shape (2802, 8184)) to: /tmp/tmp0e5ryw87/kmer_temp/kmer_binary_batches_def18359/batch_000001.npz\n",
      "  Batch 1 completed in 46.96s.\n",
      "\n",
      "Combining 1 batch matrices...\n",
      "\n",
      "\u001b[A2025-10-31 17:41:37,707 - INFO - root - Starting Loading batch matrices (1 items)\n",
      "\n",
      "\u001b[A2025-10-31 17:41:37,956 - INFO - root - Loading batch matrices: 25% complete (1/1) - elapsed: 0.3s\n",
      "2025-10-31 17:41:37,956 - INFO - root - Loading batch matrices: 50% complete (1/1) - elapsed: 0.3s\n",
      "2025-10-31 17:41:37,956 - INFO - root - Loading batch matrices: 75% complete (1/1) - elapsed: 0.3s\n",
      "\n",
      "\u001b[ACombined all batch matrices in 0.26s.\n",
      "Final sparse matrix shape: (2802, 8184)\n",
      "Processing sequences: 2802it [00:47, 59.21it/s]   \n",
      "2025-10-31 17:41:37,983 - INFO - root - Completed Processing sequences (2,802 items) - elapsed: 47.3s\n",
      "Stacking matrices: 100%|██████████| 1/1 [00:00<00:00,  3.61it/s]\n",
      "2025-10-31 17:41:37,984 - INFO - root - Completed Stacking matrices (1/1 items) - elapsed: 0.3s\n",
      "Cleaned up temporary directory: /tmp/tmp0e5ryw87/kmer_temp/kmer_binary_batches_def18359\n",
      "Attempting to create dense Pandas DataFrame (WARNING: this can be memory-intensive)...\n",
      "  Dense DataFrame created. Memory usage: 87.64 MB\n",
      "2025-10-31 17:41:38,207 - INFO - feature_wrapper - Saving K-mer results to /tmp/tmpcuq0ti_7/kmer_features.parquet\n",
      "Calculating k-mer features: 100%|██████████| 3/3 [27:06<00:00, 410.86s/modules]2025-10-31 17:41:38,927 - INFO - root - Calculating k-mer features: 75% complete (3/3) - elapsed: 1626.8s\n",
      "2025-10-31 17:41:38,927 - INFO - feature_wrapper - K-mer completed. Shape: (2802, 8184)\n",
      "2025-10-31 17:41:38,928 - INFO - feature_wrapper - Aggregating features...\n",
      "2025-10-31 17:41:38,928 - WARNING - feature_wrapper - transcript_id column not found in K-mer DataFrame. Merging may fail.\n",
      "2025-10-31 17:41:39,504 - INFO - utils.kmer_redux - Reducing 8 columns of 3-mers\n",
      "2025-10-31 17:41:39,532 - INFO - utils.kmer_redux - Reducing 16 columns of 4-mers\n",
      "2025-10-31 17:41:39,595 - INFO - utils.kmer_redux - Reducing 32 columns of 5-mers\n",
      "2025-10-31 17:41:39,661 - INFO - utils.kmer_redux - Reducing 64 columns of 6-mers\n",
      "2025-10-31 17:41:39,728 - INFO - utils.kmer_redux - Reducing 128 columns of 7-mers\n",
      "2025-10-31 17:41:39,763 - INFO - utils.kmer_redux - Reducing 256 columns of 8-mers\n",
      "2025-10-31 17:41:39,815 - INFO - utils.kmer_redux - Reducing 512 columns of 9-mers\n",
      "2025-10-31 17:41:39,889 - INFO - utils.kmer_redux - Reducing 1024 columns of 10-mers\n",
      "2025-10-31 17:41:40,029 - INFO - utils.kmer_redux - Reducing 2048 columns of 11-mers\n",
      "2025-10-31 17:41:40,191 - INFO - utils.kmer_redux - Reducing 4096 columns of 12-mers\n",
      "2025-10-31 17:41:40,358 - INFO - feature_wrapper - Dropping 5 BED coordinate columns: ['bed_chromosome', 'bed_start', 'bed_end', 'bed_score', 'bed_strand']\n",
      "2025-10-31 17:41:40,360 - INFO - feature_wrapper - Aggregated feature DataFrame shape: (2802, 44)\n",
      "Feature extraction complete: 100%|██████████| 3/3 [27:08<00:00, 542.73s/modules]\n",
      "2025-10-31 17:41:40,364 - INFO - root - Completed Feature extraction complete (3/3 items) - elapsed: 1628.2s\n"
     ]
    }
   ],
   "source": [
    "ncr = fw.run_all(\n",
    "    gtf_file='/home/chlab/flync/src/data/ncr_base.gtf',\n",
    "    config_file='/home/chlab/flync/src/config/bwq_config.yaml',\n",
    "    ref_genome_path='/home/chlab/flync/genome/Drosophila_melanogaster.BDGP6.46.dna.toplevel.fa',\n",
    "    sparse=False,\n",
    "    use_dim_redux=True,\n",
    "    redux_n_components=1,\n",
    "    use_tfidf=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c32622d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-08 08:43:33,452 - INFO - feature_wrapper - Running all feature extraction scripts...\n",
      "Feature extraction:   0%|          | 0/3 [00:00<?, ?modules/s]2025-10-08 08:43:33,456 - INFO - root - Starting Feature extraction (3 items)\n",
      "Running BigWig Query:   0%|          | 0/3 [00:00<?, ?modules/s]2025-10-08 08:43:33,460 - INFO - feature_wrapper - Running BigWig Query feature extraction...\n",
      "2025-10-08 08:43:33,460 - INFO - feature_wrapper - Using configuration file: ../config/bwq_config.yaml\n",
      "2025-10-08 08:43:33,461 - INFO - root - Loading configuration from ../config/bwq_config.yaml\n",
      "2025-10-08 08:43:33,466 - INFO - root - Loaded 26 stat/extraction configurations from ../config/bwq_config.yaml\n",
      "2025-10-08 08:43:33,467 - INFO - root - Reading genomic ranges from ../../new-tests/pcg_base.bed\n",
      "2025-10-08 08:43:33,829 - INFO - root - Extracted 23365 ranges from ../../new-tests/pcg_base.bed using pyranges.\n",
      "2025-10-08 08:43:33,830 - INFO - root - Processing queries with 20 workers\n",
      "2025-10-08 08:43:33,831 - INFO - root - Using specified max_workers: 20\n",
      "2025-10-08 08:43:33,832 - INFO - root - Preloading 10 unique file paths...\n",
      "2025-10-08 08:43:33,834 - INFO - root - Starting Downloading/caching files (10 items)\n",
      "2025-10-08 08:43:33,834 - INFO - root - Attempting to preload 10 remote files...\n",
      "2025-10-08 08:43:33,838 - INFO - root - Downloading/caching files: 25% complete (3/10) - elapsed: 0.0s\n",
      "2025-10-08 08:43:33,839 - INFO - root - Downloading/caching files: 50% complete (5/10) - elapsed: 0.0s\n",
      "2025-10-08 08:43:33,839 - INFO - root - Downloading/caching files: 75% complete (8/10) - elapsed: 0.0s\n",
      "2025-10-08 08:43:33,840 - INFO - root - Preloaded 10/10 remote files\n",
      "Downloading/caching files: 100%|██████████| 10/10 [00:00<00:00, 1489.14it/s]\n",
      "2025-10-08 08:43:33,841 - INFO - root - Completed Downloading/caching files (10/10 items) - elapsed: 0.0s\n",
      "2025-10-08 08:43:33,841 - INFO - root - Using ProcessPoolExecutor with 20 workers, batch_size=50\n",
      "2025-10-08 08:43:33,842 - INFO - root - Created 468 batches from 23365 ranges\n",
      "2025-10-08 08:43:34,019 - INFO - root - Submitted 468 batch queries to 20 workers.\n",
      "2025-10-08 08:43:34,033 - INFO - root - Starting Processing 23365 ranges in 468 batches (468 items)\n",
      "2025-10-08 08:44:13,738 - INFO - root - Processing 23365 ranges in 468 batches: 25% complete (117/468) - elapsed: 39.7s\n",
      "2025-10-08 08:44:51,565 - INFO - root - Processing 23365 ranges in 468 batches: 50% complete (234/468) - elapsed: 77.5s\n",
      "2025-10-08 08:45:34,676 - INFO - root - Processing 23365 ranges in 468 batches: 75% complete (351/468) - elapsed: 120.7s\n",
      "Processing 23365 ranges in 468 batches: 100%|██████████| 468/468 [02:50<00:00,  2.74it/s]\n",
      "2025-10-08 08:46:24,919 - INFO - root - Completed Processing 23365 ranges in 468 batches (468/468 items) - elapsed: 170.9s\n",
      "2025-10-08 08:46:24,956 - INFO - root - Collected 23365 results. Creating dataframe.\n",
      "2025-10-08 08:46:25,065 - INFO - feature_wrapper - Saving BigWig Query results to /tmp/tmp2nocx9fm/bwq_features.parquet\n",
      "Running BigWig Query:  33%|███▎      | 1/3 [02:51<05:43, 171.73s/modules]2025-10-08 08:46:25,184 - INFO - root - Running BigWig Query: 25% complete (1/3) - elapsed: 171.7s\n",
      "2025-10-08 08:46:25,185 - INFO - feature_wrapper - BigWig Query completed. Shape: (23365, 30)\n",
      "2025-10-08 08:46:25,186 - INFO - feature_wrapper - Extracting RNA sequences from BED file: ../../new-tests/pcg_base.bed\n",
      "2025-10-08 08:46:25,203 - INFO - feature_wrapper - Loaded 23365 regions from BED file\n",
      "2025-10-08 08:46:25,204 - INFO - feature_wrapper - Extracting sequences using reference genome: ../../new-tests/Drosophila_melanogaster.BDGP6.46.dna.toplevel.fa\n",
      "2025-10-08 08:46:27,820 - INFO - feature_wrapper - Writing 23365 sequences to /tmp/tmp2nocx9fm/sequences.fasta\n",
      "2025-10-08 08:46:28,585 - INFO - feature_wrapper - Successfully extracted 23365 sequences to /tmp/tmp2nocx9fm/sequences.fasta\n",
      "2025-10-08 08:46:28,592 - INFO - feature_wrapper - Sequence extraction completed. FASTA: /tmp/tmp2nocx9fm/sequences.fasta SeqDF shape: (23365, 2)\n",
      "Calculating RNA MFE:  33%|███▎      | 1/3 [02:55<05:43, 171.73s/modules] 2025-10-08 08:46:28,673 - INFO - feature_wrapper - Running MFE feature extraction...\n",
      "2025-10-08 08:46:28,676 - INFO - root - Calculating MFE for 23365 sequences using 20 processes with LinearFold...\n",
      "2025-10-08 08:46:28,678 - INFO - root - Starting Folding RNA sequences (23,365 items)\n",
      "2025-10-08 09:18:39,754 - INFO - root - Folding RNA sequences: 25% complete (5,842/23,365) - elapsed: 1931.1s\n",
      "2025-10-08 09:53:50,688 - INFO - root - Folding RNA sequences: 50% complete (11,683/23,365) - elapsed: 4042.0s\n",
      "2025-10-08 10:32:42,232 - INFO - root - Folding RNA sequences: 75% complete (17,524/23,365) - elapsed: 6373.6s\n",
      "Folding RNA sequences: 100%|██████████| 23365/23365 [2:35:11<00:00,  2.51it/s]\n",
      "2025-10-08 11:21:40,589 - INFO - root - Completed Folding RNA sequences (23,365/23,365 items) - elapsed: 9311.9s\n",
      "2025-10-08 11:21:40,593 - INFO - root - MFE calculations complete.\n",
      "2025-10-08 11:21:40,830 - INFO - feature_wrapper - Saving MFE results to /tmp/tmp2nocx9fm/mfe_features.parquet\n",
      "Calculating RNA MFE:  67%|██████▋   | 2/3 [2:38:09<1:32:31, 5551.66s/modules]2025-10-08 11:21:42,793 - INFO - root - Calculating RNA MFE: 50% complete (2/3) - elapsed: 9489.3s\n",
      "2025-10-08 11:21:42,796 - INFO - feature_wrapper - MFE completed. Shape: (23365, 4)\n",
      "Calculating k-mer features:  67%|██████▋   | 2/3 [2:38:09<1:32:31, 5551.66s/modules]2025-10-08 11:21:42,799 - INFO - feature_wrapper - Running K-mer feature extraction...\n",
      "2025-10-08 11:21:42,804 - INFO - features.kmer - Using 20 worker processes. Batch size: 10,000 sequences.\n",
      "2025-10-08 11:21:42,804 - INFO - features.kmer - Counting BINARY k-mers (A/T->0, G/C->1) for k=3-12.\n",
      "2025-10-08 11:21:42,805 - INFO - root - Generating all possible k-mers for k=3-12 using alphabet '01'...\n",
      "2025-10-08 11:21:42,808 - INFO - root - Generated 8,184 unique k-mers in 0.00s.\n",
      "Using temporary directory for batches: ../../tmp/kmer_temp/kmer_binary_batches_627b27cb\n",
      "2025-10-08 11:21:42,825 - INFO - root - Starting Processing sequences\n",
      "\n",
      "Processing Batch 1...\n",
      "Reading from file 1/1: sequences.fasta\n",
      "  Read 10000 DNA sequences for batch 1.\n",
      "  Starting binary k-mer counting for 10000 sequences in batch 1...\n",
      "  Building sparse matrix for batch 1...\n",
      "  Saving batch matrix (shape (10000, 8184)) to: ../../tmp/kmer_temp/kmer_binary_batches_627b27cb/batch_000001.npz\n",
      "  Batch 1 completed in 225.72s.\n",
      "\n",
      "Processing Batch 2...\n",
      "  Read 10000 DNA sequences for batch 2.\n",
      "  Starting binary k-mer counting for 10000 sequences in batch 2...\n",
      "  Building sparse matrix for batch 2...\n",
      "  Saving batch matrix (shape (10000, 8184)) to: ../../tmp/kmer_temp/kmer_binary_batches_627b27cb/batch_000002.npz\n",
      "  Batch 2 completed in 245.49s.\n",
      "\n",
      "Processing Batch 3...\n",
      "  Read 3365 DNA sequences for batch 3.\n",
      "  Starting binary k-mer counting for 3365 sequences in batch 3...\n",
      "  Building sparse matrix for batch 3...\n",
      "  Saving batch matrix (shape (3365, 8184)) to: ../../tmp/kmer_temp/kmer_binary_batches_627b27cb/batch_000003.npz\n",
      "  Batch 3 completed in 85.23s.\n",
      "\n",
      "Combining 3 batch matrices...\n",
      "\n",
      "\u001b[A2025-10-08 11:30:59,554 - INFO - root - Starting Loading batch matrices (3 items)\n",
      "\n",
      "\u001b[A2025-10-08 11:31:00,797 - INFO - root - Loading batch matrices: 25% complete (1/3) - elapsed: 1.2s\n",
      "\n",
      "\u001b[A2025-10-08 11:31:02,118 - INFO - root - Loading batch matrices: 50% complete (2/3) - elapsed: 2.6s\n",
      "\n",
      "\u001b[A2025-10-08 11:31:02,595 - INFO - root - Loading batch matrices: 75% complete (3/3) - elapsed: 3.0s\n",
      "\n",
      "\u001b[ACombined all batch matrices in 3.28s.\n",
      "Final sparse matrix shape: (23365, 8184)\n",
      "Processing sequences: 23365it [09:20, 41.68it/s]\n",
      "2025-10-08 11:31:03,387 - INFO - root - Completed Processing sequences (23,365 items) - elapsed: 560.6s\n",
      "Stacking matrices: 100%|██████████| 3/3 [00:03<00:00,  1.28s/it]\n",
      "2025-10-08 11:31:03,388 - INFO - root - Completed Stacking matrices (3/3 items) - elapsed: 3.8s\n",
      "Cleaned up temporary directory: ../../tmp/kmer_temp/kmer_binary_batches_627b27cb\n",
      "Attempting to create dense Pandas DataFrame (WARNING: this can be memory-intensive)...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: The final matrix dimensions (23365x8184) are large. Converting to a dense DataFrame may cause a MemoryError.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Dense DataFrame created. Memory usage: 730.78 MB\n",
      "2025-10-08 11:31:03,790 - INFO - feature_wrapper - Saving K-mer results to /tmp/tmp2nocx9fm/kmer_features.parquet\n",
      "Calculating k-mer features: 100%|██████████| 3/3 [2:47:34<00:00, 3274.69s/modules]  2025-10-08 11:31:07,896 - INFO - root - Calculating k-mer features: 75% complete (3/3) - elapsed: 10054.4s\n",
      "2025-10-08 11:31:07,896 - INFO - feature_wrapper - K-mer completed. Shape: (23365, 8184)\n",
      "2025-10-08 11:31:07,897 - INFO - feature_wrapper - Aggregating features...\n",
      "2025-10-08 11:31:07,898 - WARNING - feature_wrapper - transcript_id column not found in K-mer DataFrame. Merging may fail.\n",
      "2025-10-08 11:31:14,908 - INFO - utils.kmer_redux - Reducing 8 columns of 3-mers\n",
      "2025-10-08 11:31:15,065 - INFO - utils.kmer_redux - Reducing 16 columns of 4-mers\n",
      "2025-10-08 11:31:15,238 - INFO - utils.kmer_redux - Reducing 32 columns of 5-mers\n",
      "2025-10-08 11:31:15,436 - INFO - utils.kmer_redux - Reducing 64 columns of 6-mers\n",
      "2025-10-08 11:31:15,649 - INFO - utils.kmer_redux - Reducing 128 columns of 7-mers\n",
      "2025-10-08 11:31:15,943 - INFO - utils.kmer_redux - Reducing 256 columns of 8-mers\n",
      "2025-10-08 11:31:16,347 - INFO - utils.kmer_redux - Reducing 512 columns of 9-mers\n",
      "2025-10-08 11:31:16,903 - INFO - utils.kmer_redux - Reducing 1024 columns of 10-mers\n",
      "2025-10-08 11:31:17,801 - INFO - utils.kmer_redux - Reducing 2048 columns of 11-mers\n",
      "2025-10-08 11:31:19,126 - INFO - utils.kmer_redux - Reducing 4096 columns of 12-mers\n",
      "2025-10-08 11:31:20,984 - INFO - feature_wrapper - Dropping 5 BED coordinate columns: ['bed_chromosome', 'bed_start', 'bed_end', 'bed_score', 'bed_strand']\n",
      "2025-10-08 11:31:20,989 - INFO - feature_wrapper - Aggregated feature DataFrame shape: (23365, 44)\n",
      "Feature extraction complete: 100%|██████████| 3/3 [2:47:47<00:00, 3355.85s/modules]\n",
      "2025-10-08 11:31:20,996 - INFO - root - Completed Feature extraction complete (3/3 items) - elapsed: 10067.5s\n"
     ]
    }
   ],
   "source": [
    "pcg = fw.run_all(\n",
    "    gtf_file='/home/chlab/flync/src/data/pcg_base.gtf',\n",
    "    config_file='/home/chlab/flync/src/config/bwq_config.yaml',\n",
    "    ref_genome_path='/home/chlab/flync/genome/Drosophila_melanogaster.BDGP6.46.dna.toplevel.fa',\n",
    "    sparse=False,\n",
    "    use_dim_redux=True,\n",
    "    redux_n_components=1,\n",
    "    use_tfidf=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c91f53b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-08 12:00:39,382 - INFO - features.clean_training - Starting data preparation mode=training input_mode=paired suffix=_pyAPI scaler=standard multi_hot=True\n",
      "2025-10-08 12:00:39,394 - INFO - features.clean_training - Concatenated in-memory positive/negative DataFrames.\n",
      "2025-10-08 12:00:39,395 - INFO - features.clean_training - Label distribution stage=loaded_paired_df total_rows=26155 distribution={False: 23365, True: 2790}\n",
      "2025-10-08 12:00:39,403 - INFO - features.clean_training - Label distribution stage=pre_cleaning total_rows=26155 distribution={False: 23365, True: 2790}\n",
      "2025-10-08 12:00:39,406 - INFO - features.clean_training - Dropped non-feature columns count=5\n",
      "2025-10-08 12:00:39,408 - INFO - features.clean_training - Filled NA values with 0 column=cov_epdnew count=51\n",
      "2025-10-08 12:00:39,410 - INFO - features.clean_training - Filled NA values with 0 column=cov_h3k4me3 count=973\n",
      "2025-10-08 12:00:39,411 - INFO - features.clean_training - Filled NA values with 0 column=cov_s2_pol2 count=589\n",
      "2025-10-08 12:00:39,412 - INFO - features.clean_training - Filled NA values with 0 column=cov_tss_minus count=190\n",
      "2025-10-08 12:00:39,413 - INFO - features.clean_training - Filled NA values with 0 column=cov_tss_plus count=208\n",
      "2025-10-08 12:00:39,413 - INFO - features.clean_training - Filled NA values with 0 column=max_s2_pol2 count=589\n",
      "2025-10-08 12:00:39,414 - INFO - features.clean_training - Filled NA values with 0 column=max_tss_plus count=208\n",
      "2025-10-08 12:00:39,415 - INFO - features.clean_training - Filled NA values with 0 column=mean_h3k4me3 count=973\n",
      "2025-10-08 12:00:39,416 - INFO - features.clean_training - Filled NA values with 0 column=mean_pcons27 count=36\n",
      "2025-10-08 12:00:39,417 - INFO - features.clean_training - Filled NA values with 0 column=mean_phylocons124 count=1\n",
      "2025-10-08 12:00:39,417 - INFO - features.clean_training - Filled NA values with 0 column=min_tss_minus count=190\n",
      "2025-10-08 12:00:39,418 - INFO - features.clean_training - Filled NA values with 0 column=std__phylocons124 count=1\n",
      "2025-10-08 12:00:39,419 - INFO - features.clean_training - Filled NA values with 0 column=std_pcons27 count=36\n",
      "2025-10-08 12:00:39,420 - INFO - features.clean_training - Filled NA values with 0 column=sum__phylocons124 count=11\n",
      "2025-10-08 12:00:39,421 - INFO - features.clean_training - Filled NA values with 0 column=sum_h3k4me3 count=2054\n",
      "2025-10-08 12:00:39,421 - INFO - features.clean_training - Filled NA values with 0 column=sum_pcons27 count=62\n",
      "2025-10-08 12:00:39,422 - INFO - features.clean_training - Filled NA values with 0 column=sum_s2_pol2 count=656\n",
      "2025-10-08 12:00:39,423 - INFO - features.clean_training - Filled NA values with 0 column=sum_tss_minus count=278\n",
      "2025-10-08 12:00:39,424 - INFO - features.clean_training - Filled NA values with 0 column=sum_tss_plus count=276\n",
      "2025-10-08 12:00:39,433 - INFO - features.clean_training - Filled NA values with empty string column=entries_remap count=217\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-08 12:00:39,538 - INFO - features.clean_training - Label distribution stage=post_cleaning total_rows=26155 distribution={False: 23365, True: 2790}\n",
      "2025-10-08 12:01:04,488 - INFO - features.clean_training - Applied multi-hot encoding processed_columns=2 generated_features=695\n",
      "2025-10-08 12:01:04,526 - INFO - features.clean_training - Scaled numeric features scaler=standard feature_count=36\n",
      "2025-10-08 12:01:04,573 - INFO - features.clean_training - Split summary split=train rows=18308 distribution={False: 0.8933, True: 0.1067}\n",
      "2025-10-08 12:01:04,583 - INFO - features.clean_training - Split summary split=val rows=3923 distribution={False: 0.8934, True: 0.1066}\n",
      "2025-10-08 12:01:04,584 - INFO - features.clean_training - Split summary split=test rows=3924 distribution={False: 0.8932, True: 0.1068}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/chlab/flync/src/features/clean_training.py:839: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[-0.30050981 -0.31903434 -0.34313423 ... -0.35864628 -0.35446477\n",
      "  4.80880037]' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  df.loc[:, numeric_cols] = scaler.fit_transform(df[numeric_cols])\n",
      "/home/chlab/flync/src/features/clean_training.py:839: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[-1.57199824 -2.0266666  -0.82179545 ... -1.776599   -1.776599\n",
      "  1.51974658]' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.\n",
      "  df.loc[:, numeric_cols] = scaler.fit_transform(df[numeric_cols])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-10-08 12:01:04,803 - INFO - features.clean_training - Saved dataframe path=../../cleaned_training_data.parquet rows=26155 cols=732\n",
      "2025-10-08 12:01:04,979 - INFO - features.clean_training - Saved split split=train path=../../X_train__pyAPI.parquet rows=18308 cols=732\n",
      "2025-10-08 12:01:05,071 - INFO - features.clean_training - Saved split split=val path=../../X_val__pyAPI.parquet rows=3923 cols=732\n",
      "2025-10-08 12:01:05,162 - INFO - features.clean_training - Saved split split=test path=../../X_test__pyAPI.parquet rows=3924 cols=732\n",
      "2025-10-08 12:01:05,163 - INFO - features.clean_training - Pipeline finished mode=training cleaned_path=../../cleaned_training_data.parquet split_paths={'train': '../../X_train__pyAPI.parquet', 'val': '../../X_val__pyAPI.parquet', 'test': '../../X_test__pyAPI.parquet'}\n"
     ]
    }
   ],
   "source": [
    "plc = ct.PipelineConfig(\n",
    "    positive_df=ncr,\n",
    "    negative_df=pcg,\n",
    "    label_column='y',\n",
    "    train_size=0.7,\n",
    "    test_size=0.15,\n",
    "    val_size=0.15,\n",
    "    split_output_dir='/home/chlab/flync/data',\n",
    "    cleaned_output_path='/home/chlab/flync/data/cleaned_training_data.parquet',\n",
    "    random_seed=99,\n",
    "    scaler='standard',\n",
    "    scaler_path='/home/chlab/flync/models/scaler.pkl',\n",
    "    enable_multi_hot=True,\n",
    "    mode='training',\n",
    ")\n",
    "\n",
    "res = ct.prepare_training_data(plc)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "flync",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
